File: app/src/main/java/com/johnnyzhou/movieme/ui/movie/list/MovieListPresenter.java
package com.johnnyzhou.movieme.ui.movie.list;

import com.johnnyzhou.movieme.di.module.AppModule;
import com.johnnyzhou.movieme.ui.movie.Movie;
import com.johnnyzhou.movieme.util.NetworkUtil;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Named;

import rx.Scheduler;
import rx.functions.Action1;
import rx.subscriptions.CompositeSubscription;

public class MovieListPresenter implements MovieListContract.Presenter<MovieListContract.View> {
    private Map<String, List<Movie>> currentQueryMap;
    private MovieListInteractor interactor;
    private CompositeSubscription subscriptions;
    private MovieListContract.View view;
    private List<Movie> popularMovies;
    private Scheduler mainThread;

    private static MovieListContract.View emptyView = new MovieListContract.View() {
        @Override
        public void showMovies(List<Movie> movies) { // Do nothing
        }

        @Override
        public void showLoading(boolean show) {
            // Do nothing
        }

        @Override
        public void showError(String message) {
            // Do nothing
        }
    };

    @Inject
    public MovieListPresenter(MovieListInteractor interactor,
                              @Named(AppModule.MAIN_THREAD) Scheduler mainThread) {

        subscriptions = new CompositeSubscription();
        this.interactor = interactor;
        this.mainThread = mainThread;
        currentQueryMap = new HashMap<>();
        view = emptyView;
    }

    @Override
    public void searchMovie(final String query) {
        showLoading();

        if (currentQueryMap.containsKey(query)) {
            showMovies(currentQueryMap.get(query));
            return;
        }

        currentQueryMap.clear();
        interactor.searchMovies(query)
                .observeOn(mainThread)
                .subscribe(new Action1<List<Movie>>() {
                    @Override
                    public void call(List<Movie> searchResult) {
                        currentQueryMap.put(query, searchResult);
                        showMovies(searchResult);
                    }
                }, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable throwable) {
                        showError(NetworkUtil.NETWORK_ERROR_MSG);
                    }
                });
    }

    @Override
    public void getPopularMovies() {
        showLoading();

        if (popularMovies != null) {
            showMovies(popularMovies);
            return;
        }

        Action1<List<Movie>> onNext = new Action1<List<Movie>>() {
            @Override
            public void call(List<Movie> movies) {
                MovieListPresenter.this.popularMovies = movies;
                showMovies(movies);
            }
        };

        Action1<Throwable> onError = new Action1<Throwable>() {
            @Override
            public void call(Throwable e) {
                showError(NetworkUtil.NETWORK_ERROR_MSG);
            }
        };

        subscriptions.add(interactor.getPopularMovies()
                .observeOn(mainThread)
                .subscribe(onNext, onError));
    }

    private void showMovies(final List<Movie> movies) {
        view.showLoading(false);
        view.showMovies(movies);
    }

    private void showLoading() {
        view.showError(null);
        view.showLoading(true);
    }

    private void showError(String errorMsg) {
        view.showLoading(false);
        view.showError(errorMsg);
    }

    @Override
    public void bindView(MovieListContract.View view) {
        if (subscriptions.isUnsubscribed())
            subscriptions = new CompositeSubscription();

        this.view = view;
    }

    @Override
    public void unbindView() {
        view = emptyView;
    }

    @Override
    public void onDestroy() {
        subscriptions.clear();
    }
}

File: app/src/main/java/com/johnnyzhou/movieme/ui/movie/list/MovieListPresenter.kt
package com.johnnyzhou.movieme.ui.movie.list

import com.johnnyzhou.movieme.di.module.AppModule
import com.johnnyzhou.movieme.ui.movie.Movie
import com.johnnyzhou.movieme.util.NetworkUtil
import rx.Scheduler
import rx.functions.Action1
import rx.subscriptions.CompositeSubscription
import java.util.*
import javax.inject.Inject
import javax.inject.Named

//class MovieListPresenter @Inject constructor(
//    interactor: MovieListInteractor,
//    @Named(AppModule.MAIN_THREAD) mainThread: Scheduler
//) : MovieListContract.Presenter<MovieListContract.View?> {
//    private val currentQueryMap: MutableMap<String, List<Movie>>
//    private val interactor: MovieListInteractor
//    private var subscriptions: CompositeSubscription
//    private var view: MovieListContract.View
//    private var popularMovies: List<Movie>? = null
//    private val mainThread: Scheduler
//    override fun searchMovie(query: String) {
//        showLoading()
//        if (currentQueryMap.containsKey(query)) {
//            showMovies(currentQueryMap[query]!!)
//            return
//        }
//        currentQueryMap.clear()
//        interactor.searchMovies(query)
//            .observeOn(mainThread)
//            .subscribe({ searchResult ->
//                currentQueryMap[query] = searchResult
//                showMovies(searchResult)
//            }) { showError(NetworkUtil.NETWORK_ERROR_MSG) }
//    }
//
//    override fun getPopularMovies() {
//        showLoading()
//        if (popularMovies != null) {
//            showMovies(popularMovies!!)
//            return
//        }
//        val onNext =
//            Action1<List<Movie>> { movies ->
//                popularMovies = movies
//                showMovies(movies)
//            }
//        val onError: Action1<Throwable> = Action1 { showError(NetworkUtil.NETWORK_ERROR_MSG) }
//        subscriptions.add(
//            interactor.popularMovies
//                .observeOn(mainThread)
//                .subscribe(onNext, onError)
//        )
//    }
//
//    private fun showMovies(movies: List<Movie>) {
//        view.showLoading(false)
//        view.showMovies(movies)
//    }
//
//    private fun showLoading() {
//        view.showError(null)
//        view.showLoading(true)
//    }
//
//    private fun showError(errorMsg: String) {
//        view.showLoading(false)
//        view.showError(errorMsg)
//    }
//
//    override fun bindView(view: MovieListContract.View) {
//        if (subscriptions.isUnsubscribed) subscriptions = CompositeSubscription()
//        this.view = view
//    }
//
//    override fun unbindView() {
//        view = emptyView
//    }
//
//    override fun onDestroy() {
//        subscriptions.clear()
//    }
//
//    companion object {
//        private val emptyView: MovieListContract.View = object : MovieListContract.View {
//            override fun showMovies(movies: List<Movie>) { // Do nothing
//            }
//
//            override fun showLoading(show: Boolean) {
//                // Do nothing
//            }
//
//            override fun showError(message: String) {
//                // Do nothing
//            }
//        }
//    }
//
//    init {
//        subscriptions = CompositeSubscription()
//        this.interactor = interactor
//        this.mainThread = mainThread
//        currentQueryMap = HashMap()
//        view = emptyView
//    }
//}